book-content/chapters/09-typescript-only-features.md

```diff
@@ -0,0 +1,518 @@
+
+The result would then be similar to the following:
+
+```javascript
+var AlbumStatus = {
+  0: "NewRelease",
+  1: "OnSale",
+  2: "StaffPick",
+  NewRelease: 0,
+  OnSale: 1,
+  StaffPick: 2,
+};
+```
+
+This reverse mapping means that there are more keys available on an enum than you might expect. So, performing an `Object.keys` call on an enum will return both the keys and the values.
+
+```typescript
+console.log(Object.keys(AlbumStatus)); // ["0", "1", "2", "NewRelease", "OnSale", "StaffPick"]
+```
+
+This can be a real gotcha if you're not expecting it.
+
+#### How String Enums Transpile
+
+String enums don't have the same behavior as numeric enums. When you specify string values, the transpiled JavaScript is much simpler:
+
+```typescript
+enum AlbumStatus {
+  NewRelease = "NEW_RELEASE",
+  OnSale = "ON_SALE",
+  StaffPick = "STAFF_PICK",
+}
+```
+
+```javascript
+var AlbumStatus;
+(function (AlbumStatus) {
+  AlbumStatus["NewRelease"] = "NEW_RELEASE";
+  AlbumStatus["OnSale"] = "ON_SALE";
+  AlbumStatus["StaffPick"] = "STAFF_PICK";
+})(AlbumStatus || (AlbumStatus = {}));
+```
+
+Now, there is no reverse mapping, and the object only contains the enum values. An `Object.keys` call will only return the keys, as you might expect.
+
+```typescript
+console.log(Object.keys(AlbumStatus)); // ["NewRelease", "OnSale", "StaffPick"]
+```
+
+This difference between numeric and string enums feels inconsistent, and it can be a source of confusion.
+
+#### Numeric Enums Behave Like Union Types
+
+Another odd feature of enums is that string enums and numeric enums behave differently when used as types.
+
+Let's redefine our `logStatus` function with a numeric enum:
+
+```typescript
+enum AlbumStatus {
+  NewRelease = 0,
+  OnSale = 1,
+  StaffPick = 2,
+}
+
+function logStatus(genre: AlbumStatus) {
+  console.log(genre);
+}
+```
+
+Now, we can call `logStatus` with a member of the enum:
+
+```typescript
+logStatus(AlbumStatus.NewRelease);
+```
+
+But we can also call it with a plain number:
+
+```typescript
+logStatus(0);
+```
+
+If we call it with a number that isn't a member of the enum, TypeScript will report an error:
+
+```ts twoslash
+// @errors: 2345
+enum AlbumStatus {
+  NewRelease = 0,
+  OnSale = 1,
+  StaffPick = 2,
+}
+
+function logStatus(genre: AlbumStatus) {
+  console.log(genre);
+}
+
+// ---cut---
+logStatus(3);
+```
+
+This is different from string enums, which only allow the enum members to be used as types:
+
+```ts twoslash
+// @errors: 2345
+enum AlbumStatus {
+  NewRelease = "NEW_RELEASE",
+  OnSale = "ON_SALE",
+  StaffPick = "STAFF_PICK",
+}
+
+function logStatus(genre: AlbumStatus) {
+  console.log(genre);
+}
+
+logStatus(AlbumStatus.NewRelease);
+logStatus("NEW_RELEASE");
+```
+
+The way string enums behave feels more natural - it matches how enums work in other languages like C# and Java.
+
+But the fact that they're not consistent with numeric enums can be a source of confusion.
```

Comment:
It also came to my attention recently that numeric enums are just not type-safe... to support usage of bitmasks

https://www.typescriptlang.org/play/?#code/MYewdgzgLgBApmArgWxgWQJ4FEmoN4BQMMAggDREwBCFxAwjALykwBk1FAvgQaJLAhQwA8lAAWcAE44hhYuUo1KDZiTYcC3AgDNEYYFACW4GFDjQAFMgCGEANYAudNlwBKGHi1nLARgAMrgDcMAD0IfCSkiCSZDCgIAA2BAAmcMAJ1pJwceDQMLhOuABGUoEE3lAWbsFh+SARUTF1sPFJ5eaVohLSuAB0JEGh4VKNsa3tlpgyyP0wAD4i4lLTvVSDtWD1I9Gxmy0giQRAA